### PR TITLE
Singleton service

### PR DESCRIPTION
**Pull request by tardyp**
This class implements a generic Service that needs to be instantiated as a singleton according to its parameters.
It is a common use case to need this for accessing remote services.
Having a singleton to access a remote allows to limit the number of simultaneous access to the same services.
Thus, several completely independent buildbot services can use that singleton to access the service, and automatically synchronize themselves to not overwhelm it.

### CODE REVIEW

File: master/buildbot/util/service.py [hunk 1]
```diff
@@ -109,6 +114,42 @@ def master(self):
...
+class SingletonService(AsyncMultiService):
+    """a service that is created only once per parameter set in a parent service"""
+
+    @classmethod
+    def getService(cls, parent, *args, **kwargs):
+        name = cls.getName(*args, **kwargs)
+        if name in parent.namedServices:
+            return defer.succeed(parent.namedServices[name])
+        try:
+            o = cls(*args, **kwargs)
```
Comment (sa2ajj) [inline master/buildbot/util/service.py]: o? oh? oh, no!

File: master/buildbot/util/service.py [hunk 2]
```diff
@@ -109,6 +114,42 @@ def master(self):
...
+            # This make sense as the singleton service is used as a dependency
+            # for other service
+            parent.services.remove(o)
+            parent.services.insert(0, o)
+            return o
+        return d
+
+    @classmethod
+    def getName(cls, *args, **kwargs):
+        h = hashlib.sha1()
```
Comment (sa2ajj) [inline master/buildbot/util/service.py]: this PR seems to love one letter names

File: master/buildbot/worker/hyper.py [hunk 3]
```diff
@@ -41,16 +42,49 @@
...
+    This manager manages its own thread pull, as Hyper_sh is blocking
+    """
+    name = "HyperLatentManager"
+    MAX_THREADS = 5
+
+    def __init__(self, hyper_host, hyper_accesskey, hyper_secretkey):
+        service.SingletonService.__init__(self)
+        self.name = self.getName(hyper_host, hyper_accesskey, hyper_secretkey)
+        # Prepare the parameters for the Docker Client object.
+        self.client_args = {'clouds': {
```
Comment (sa2ajj) [inline master/buildbot/worker/hyper.py]: do we need to have `client_args` public?

File: master/buildbot/worker/hyper.py [hunk 4]
```diff
@@ -41,16 +42,49 @@
...
+            hyper_host: {
+                "accesskey": hyper_accesskey,
+                "secretkey": hyper_secretkey
+            }
+        }}
+
+    def startService(self):
+        self.threadPool = threadpool.ThreadPool(
```
Comment (sa2ajj) [inline master/buildbot/worker/hyper.py]: ditto

File: master/buildbot/worker/hyper.py [hunk 5]
```diff
@@ -41,16 +42,49 @@
...
+            minthreads=1, maxthreads=self.MAX_THREADS, name='hyper')
+        self.threadPool.start()
+        self.client = Hyper(self.client_args)
```
Comment (sa2ajj) [inline master/buildbot/worker/hyper.py]: ditto

File: master/docs/developer/utils.rst [hunk 6]
```diff
@@ -886,6 +886,44 @@ For example, a particular daily scheduler could be configured on multiple master
         Therefore, in this method it is safe to reassign the "active" status to another instance.
         This method may return a Deferred.
 
+.. py:class:: SingletonService
+
+    This class implements a generic Service that needs to be instantiated as a singleton according to its parameters.
+    It is a common use case to need this for accessing remote services.
+    Having a singleton to access a remote allows to limit the number of simultaneous access to the same services.
+    Thus, several completely independent buildbot services can use that singleton to access the service, and automatically synchronize themselves to not overwhelm it.
```
Comment (sa2ajj) [inline master/docs/developer/utils.rst]: Buildbot

File: master/docs/developer/utils.rst [hunk 7]
```diff
@@ -886,6 +886,44 @@ For example, a particular daily scheduler could be configured on multiple master
...
+        This means that the SingletonService will not be destroyed when there is no more users, it will be destroyed at the master's stopService
+
+        The lifecycle of the SingletonService is the same as a service, it must implement startService and stopService in order to allocate and free its resources.
+
+    .. py:method:: getName(cls, *args, **kwargs)
+
+        Class method.
+        Takes same arguments as the constructor of the service.
+        Get a unique name for that instance of a service.
+        Used to decide if we need to create a new object.
```
Comment (sa2ajj) [inline master/docs/developer/utils.rst]: I am not sure I understand what this means.

File: master/docs/developer/utils.rst [hunk 8]
```diff
@@ -886,6 +886,44 @@ For example, a particular daily scheduler could be configured on multiple master
...
+    .. py:method:: getService(cls, parentService, *args, **kwargs)
+
+        :param parentService: an AsyncMultiService class where to lookup and register the singleton (usually the root service, the master)
+        :returns: instance of the service via Deferred
+
+        Class method.
+        Takes same arguments as the constructor of the service (plus the `parentService` at the beginning of the list).
+        Construct an instance of the service if needed, and place it at the beginning of the `parentService` service list.
+        Placing it at the beginning will guarantee that the singleton will be stopped after the other services.
+
```
Comment (sa2ajj) [inline master/docs/developer/utils.rst]: one empty line here will be enough.

File: master/buildbot/util/service.py [hunk 9]
```diff
@@ -1,4 +1,3 @@
-# This file is part of Buildbot.  Buildbot is free software: you can
```
Comment (sa2ajj) [inline master/buildbot/util/service.py]: this line is probably removed by accident

File: master/buildbot/util/service.py [hunk 10]
```diff
@@ -109,6 +114,42 @@ def master(self):
...
+    @classmethod
+    def getService(cls, parent, *args, **kwargs):
+        name = cls.getName(*args, **kwargs)
+        if name in parent.namedServices:
+            return defer.succeed(parent.namedServices[name])
+        try:
+            o = cls(*args, **kwargs)
+        except Exception:
+            return defer.fail(failure.Failure())
+        o.name = name
```
Comment (sa2ajj) [inline master/buildbot/util/service.py]: I'm not sure I like this forceful setting of `.name`

File: master/buildbot/util/service.py (hunk 2 above)
Comment (tardyp) [inline master/buildbot/util/service.py]: :) h sounds like hash which is a reserved word. 
You are right

File: master/docs/developer/utils.rst [hunk 11]
```diff
@@ -886,6 +886,44 @@ For example, a particular daily scheduler could be configured on multiple master
         Therefore, in this method it is safe to reassign the "active" status to another instance.
         This method may return a Deferred.
 
+.. py:class:: SingletonService
+
+    This class implements a generic Service that needs to be instantiated as a singleton according to its parameters.
```
Comment (Stibbons) [inline master/docs/developer/utils.rst]: if you can have n instances of this service but with different parameters this is not a singleton... for accessing to external services, we use normal services "mounted" into a known proxy (with an hardcoded name, which is not very elegant), for instance to be able to have 1 gerrit service talking to Gerrit 2.6 and another talking to Gerrit 2.11.

Comment (Stibbons) [inline master/docs/developer/utils.rst]: ok, but I advice to be very careful in defining it in the class doc for programmers. Of course, users do not care about this detail. But, for most programmers, Singleton "is a design pattern that restricts the instantiation of a class to one object", "An implementation of the singleton pattern must: ensure that only one instance of the singleton class ever exists; and provide global access to that instance." [wikipedia](https://en.wikipedia.org/wiki/Singleton_pattern)

Comment (tardyp) [inline master/docs/developer/utils.rst]: I think you are right, we need to change the name
http://stackoverflow.com/questions/1050991/singleton-with-arguments-in-java

File: master/buildbot/util/service.py [hunk 12]
```diff
@@ -110,6 +116,46 @@ def master(self):
...
+    """a service that is created only once per parameter set in a parent service"""
+
+    @classmethod
+    def getService(cls, parent, *args, **kwargs):
+        name = cls.getName(*args, **kwargs)
+        if name in parent.namedServices:
+            return defer.succeed(parent.namedServices[name])
+        try:
+            instance = cls(*args, **kwargs)
+        except Exception:
```
Comment (sa2ajj) [inline master/buildbot/util/service.py]: Do we really want catch all `Exception`s here?

File: master/buildbot/worker/hyper.py [hunk 13]
```diff
@@ -41,17 +42,50 @@
 log = Logger()
 
 
+class HyperLatentManager(service.SharedService):
+    """A shared service class that manages all the connections to the hyper cloud
+
+    There is one instance of this manager per host, accesskey, secretkey tuple.
+    This manager manages its own thread pull, as Hyper_sh is blocking
+    """
+    name = "HyperLatentManager"
```
Comment (sa2ajj) [inline master/buildbot/worker/hyper.py]: quotes

File: master/buildbot/worker/hyper.py [hunk 14]
```diff
@@ -41,17 +42,50 @@
...
+    MAX_THREADS = 5
```
Comment (sa2ajj) [inline master/buildbot/worker/hyper.py]: do we use this value outside?  if not `_MAX_THREADS`

Comment (tardyp) [inline master/buildbot/worker/hyper.py]: yes, this is eventually designed to be overriden

**Comment by tardyp** [comment]
@sa2ajj I think I answered to all the reviews, I am going to merge this as soon as travis is done.

**Comment by sa2ajj** [comment]
@tardyp do you really want us to review the change?

**Comment by tardyp** [comment]
@sa2ajj I do want reviews. I think I addressed everything, and I wanted to move forward, as I have several dependent branches.
If you have anymore issue, I'd be happy to address them in a follow up PR

**Comment by sa2ajj** [comment]
You substantially changed things since the original PR, if I had been the author, I'd wait for the other review, not just Travis results.

**Comment by tardyp** [comment]
@sa2ajj I just added two trivial comments since your last review, and a note in relnotes
https://github.com/buildbot/buildbot/pull/2444/commits/5ad63263b4ff70dc7465509c3adab298fadfe8cd
https://github.com/buildbot/buildbot/pull/2444/commits/e3c6b0570b2b04882bf953e62aba352d22abe740

There are 800loc to be reviewed in #2446 which are harder to review if the dependent branch is not merged.

I am really sorry if I upset you, this was really not my intention.

I'm just trying to use the sparse review time the best as I can so that we can move as fast as we can.

I have a lot of time now to develop buildbot, and I won't in a few weeks/months, when I'll need to transition to my new paid job (you know how it is).

**Pull request merged**